gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java

```diff
@@ -0,0 +1,468 @@
+
+    @Override
+    public String getMessageType() {
+      return Message.MessageType.SHUTDOWN.toString();
+    }
+
+    @Override
+    public void reset() {
+
+    }
+
+    /**
+     * A custom {@link MessageHandler} for handling messages of sub type
+     * {@link HelixMessageSubTypes#APPLICATION_MASTER_SHUTDOWN}.
+     */
+    private class ControllerShutdownMessageHandler extends MessageHandler {
+
+      public ControllerShutdownMessageHandler(Message message, NotificationContext context) {
+        super(message, context);
+      }
+
+      @Override
+      public HelixTaskResult handleMessage() throws InterruptedException {
+        String messageSubType = this._message.getMsgSubType();
+        Preconditions.checkArgument(
+            messageSubType.equalsIgnoreCase(HelixMessageSubTypes.APPLICATION_MASTER_SHUTDOWN.toString()),
+            String.format("Unknown %s message subtype: %s", Message.MessageType.SHUTDOWN.toString(), messageSubType));
+
+        HelixTaskResult result = new HelixTaskResult();
+
+        if (stopInProgress) {
+          result.setSuccess(true);
+          return result;
+        }
+
+        LOGGER.info("Handling message " + HelixMessageSubTypes.APPLICATION_MASTER_SHUTDOWN.toString());
+
+        ScheduledExecutorService shutdownMessageHandlingCompletionWatcher =
+            MoreExecutors.getExitingScheduledExecutorService(new ScheduledThreadPoolExecutor(1));
+
+        // Schedule the task for watching on the removal of the shutdown message, which indicates that
+        // the message has been successfully processed and it's safe to disconnect the HelixManager.
+        shutdownMessageHandlingCompletionWatcher.scheduleAtFixedRate(new Runnable() {
+          @Override
+          public void run() {
+            HelixManager helixManager = _notificationContext.getManager();
+            HelixDataAccessor helixDataAccessor = helixManager.getHelixDataAccessor();
+
+            HelixProperty helixProperty = helixDataAccessor
+                .getProperty(_message.getKey(helixDataAccessor.keyBuilder(), helixManager.getInstanceName()));
+            // The absence of the shutdown message indicates it has been removed
+            if (helixProperty == null) {
+              eventBus.post(new ApplicationMasterShutdownRequest());
+            }
+          }
+        }, 0, 1, TimeUnit.SECONDS);
+
+        result.setSuccess(true);
+        return result;
+      }
+
+      @Override
+      public void onError(Exception e, ErrorCode code, ErrorType type) {
+        LOGGER.error(
+            String.format("Failed to handle message with exception %s, error code %s, error type %s", e, code, type));
+      }
+    }
+  }
+
+  /**
+   * A custom {@link MessageHandlerFactory} for {@link ControllerUserDefinedMessageHandler}s that
+   * handle messages of type {@link org.apache.helix.model.Message.MessageType#USER_DEFINE_MSG}.
+   */
+  private class ControllerUserDefinedMessageHandlerFactory implements MessageHandlerFactory {
```

Comment:
This seems to be identical to the equivalent class in `GobblinWorkUnitRunner.java`. But I'm guessing we didn't merge the two in case other user defined messages need to be added. Would it be easier to use something like protobuf for this kind of communication. Helix doesn't seem to provide a generic and easy way to do this. Not a pressing issue, but may be useful down the line, especially when the containers + app master need to communicate more.